vllm/worker/model_runner.py

```diff
@@ -845,13 +921,67 @@ def execute_model(
         if not sampling_metadata.perform_sampling:
             return None
 
+        if self.is_mamba:
+            for i, offset in enumerate(indices):
+                self.mamba_cache[0][:, offset].copy_(conv_state[:, i])
+                self.mamba_cache[1][:, offset].copy_(ssm_state[:, i])
+
         # Sample the next token.
         output = self.model.sample(
             logits=logits,
             sampling_metadata=sampling_metadata,
         )
         return output
 
+    def _get_first_free_mamba_cache_index(self):
+        max_possible_bs = self.mamba_cache[0].shape[1]
+        occupied = [
+            id for seq_ids in self.request2i.values()
+            for id in seq_ids.values()
+        ]
+        first_free_index = [i not in occupied
+                            for i in range(max_possible_bs)].index(True)
+        return first_free_index
+
+    def _prepare_request_mamba_cache(self, requests_info: List[RequestInfo],
+                                     batch_size: int):
+        indices = []
+        max_possible_bs = self.mamba_cache[0].shape[1]
+        for request_info in requests_info:
+            cur_rid = request_info.request_id
+            if cur_rid not in self.request2i:
+                self.request2i[cur_rid] = {}
+                for seq_id in request_info.seqs_id:
+                    f_free_index = self._get_first_free_mamba_cache_index()
+                    self.request2i[cur_rid][seq_id] = f_free_index
+                    indices.append(f_free_index)
+            else:
+                for seq_id in request_info.seqs_id:
+                    if seq_id not in self.request2i[cur_rid]:
+                        f_free_index = self._get_first_free_mamba_cache_index()
+                        ## case of decoding n>1
+                        i_exist = list(self.request2i[cur_rid].values())[0]
+                        self.mamba_cache[0][:, f_free_index].copy_(
+                            self.mamba_cache[0][:, i_exist])
+                        self.mamba_cache[1][:, f_free_index].copy_(
+                            self.mamba_cache[1][:, i_exist])
+                        self.request2i[cur_rid][seq_id] = f_free_index
+                    indices.append(self.request2i[cur_rid][seq_id])
+        ## Pad the batch incase of running batch that was not captured via CG
+        padded_indices = indices
+        for _ in range(batch_size - len(indices)):
+            occu = [
+                i for s_ids in self.request2i.values() for i in s_ids.values()
+            ]
+            padded_indices += [[
+                i not in set(occu).union(padded_indices)
+                for i in range(max_possible_bs)
+            ].index(True)]
+
+        conv_state = self.mamba_cache[0][:, padded_indices]
+        ssm_state = self.mamba_cache[1][:, padded_indices]
```

Comment:
If we perform indexing here, why do we still need the following copy on line 964-957?

```
                        self.mamba_cache[0][:, f_free_index].copy_(
                            self.mamba_cache[0][:, i_exist])
                        self.mamba_cache[1][:, f_free_index].copy_(
                            self.mamba_cache[1][:, i_exist])
```

Reply:
Lines 964-957 are used to duplicate the mamba cache in case of decoding with n>1 (similar to CacheEngine.copy()) , Lines 981-982 are actually responsible for copying the mamba cache after the forward pass,  , from the `mamba_cache4gc` "buffer"  back into the right place in mamba cache. `mamba_cache4gc` acts as a fixed size memory buffer that his sole purpose is to inhabit the mamba cache during the decoding process. We cannot use the same buffer `mamba_cache` both as the fixed size memory for GC and the persistent MambaCache since it will override existing data. Ideally, It would be best if we could have a "PagedMamba" implementation where we would not need to do those exchanges in the memory.